hls-test-utils/src/Test/Hls.hs

```diff
@@ -9,38 +10,127 @@ module Test.Hls
+import           Development.Shake                 (ShakeOptions (shakeThreads))
+import           GHC.IO.Handle
+import           Ide.Plugin.Config                 (Config, formattingProvider)
+import           Ide.PluginUtils                   (pluginDescToIdePlugins)
+import           Ide.Types
 import           Language.LSP.Test
 import           Language.LSP.Types
+import           Language.LSP.Types.Capabilities   (ClientCapabilities)
+import           System.Directory                  (getCurrentDirectory,
+                                                    setCurrentDirectory)
+import           System.IO.Extra
+import           System.IO.Unsafe                  (unsafePerformIO)
+import           System.Process.Extra              (createPipe)
+import           System.Time.Extra
 import           Test.Hls.Util
-import           Test.Tasty                      hiding (Timeout)
+import           Test.Tasty                        hiding (Timeout)
 import           Test.Tasty.ExpectedFailure
 import           Test.Tasty.Golden
 import           Test.Tasty.HUnit
 import           Test.Tasty.Ingredients.Rerun
-import           Test.Tasty.Runners
-import           Test.Tasty.Runners.AntXML
 
--- | ingredient: xml runner writes json file of test results (https://github.com/ocharles/tasty-ant-xml/blob/master/Test/Tasty/Runners/AntXML.hs)
---               rerunningTests allow rerun of failed tests (https://github.com/ocharles/tasty-rerun/blob/master/src/Test/Tasty/Ingredients/Rerun.hs)
+-- | Run 'defaultMainWithRerun', limiting each single test case running at most 10 minutes
 defaultTestRunner :: TestTree -> IO ()
-defaultTestRunner =
-  defaultMainWithIngredients
-    [antXMLRunner, rerunningTests [listingTests, consoleTestReporter]]
+defaultTestRunner = defaultMainWithRerun . adjustOption (const $ mkTimeout 600000000)
 
 gitDiff :: FilePath -> FilePath -> [String]
 gitDiff fRef fNew = ["git", "diff", "--no-index", "--text", "--exit-code", fRef, fNew]
 
 goldenGitDiff :: TestName -> FilePath -> IO ByteString -> TestTree
 goldenGitDiff name = goldenVsStringDiff name gitDiff
 
-testCommand :: String
-testCommand = "test-server"
+runSessionWithServer :: PluginDescriptor IdeState -> FilePath -> Session a -> IO a
+runSessionWithServer plugin = runSessionWithServer' [plugin] def def fullCaps
+
+runSessionWithServerFormatter :: PluginDescriptor IdeState -> String -> FilePath -> Session a -> IO a
+runSessionWithServerFormatter plugin formatter =
+  runSessionWithServer'
+    [plugin]
+    def {formattingProvider = T.pack formatter}
+    def
+    fullCaps
+
+-- | Run an action, with stderr silenced
+silenceStderr :: IO a -> IO a
+silenceStderr action = withTempFile $ \temp ->
+  bracket (openFile temp ReadWriteMode) hClose $ \h -> do
+    old <- hDuplicate stderr
+    buf <- hGetBuffering stderr
+    h `hDuplicateTo'` stderr
+    action `finally` do
+      old `hDuplicateTo'` stderr
+      hSetBuffering stderr buf
+      hClose old
```

Comment:
Does this mean that we can no longer use `LSP_TEST_LOG_STDERR=1` to get more info about a failing test, since silencing stderr is now hard coded?

Reply:
//cc @berberman

Reply:
When I attempted this for all the ghcide testsuite, I installed a logger that sends output to the LSP channel. 

https://github.com/haskell/haskell-language-server/pull/1752/files?file-filters%5B%5D=.hs&file-filters%5B%5D=.project#diff-46837fe41900682e7ad33f73eee0194cde2e864e3aebd7775d412660b61029fcR1

Reply:
We could probably check for the environment variable manually and pretend that it was lsp-test that did it, to minimize disturbance of people's workflows.

(Here's lsp-test's code for it: https://github.com/haskell/lsp/blob/e707cbf5ca7077f70884ae0d2a8d016aa30ced5a/lsp-test/src/Language/LSP/Test.hs#L267-L275)

That way the current uses, like in [ci](https://github.com/haskell/haskell-language-server/blob/master/.github/workflows/test.yml#L161) doesn't have to change. Regardless of what we do, it should probably be documented together with the `LSP_TEST_LOG_MESSAGES` env variable in the [contributing](https://github.com/haskell/haskell-language-server/blob/master/docs/contributing/contributing.md) section, which it is currently not.

Reply:
agree, pr's welcome!

Reply:
> Does this mean that we can no longer use `LSP_TEST_LOG_STDERR=1` to get more info about a failing test, since silencing stderr is now hard coded?

Yes, output from lsp server is completely removed in test suites which use this function since that change, because there were too many logs were printed not through the logger, but were directly written to stderr, messing up testing status. So setting environment variable LSP_TEST_LOG_MESSAGES actually won't work. However, this function is a temporary workaround, and we should cleanup such code in server.